[EC-599] Access Selector Component

shane-melton opened the pull request:
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Adds a re-useable Web component for managing collection/groups/member permissions and access for the OAVR. 

See the added storybook stories for usage details and examples.

This cherry picks the `FormSelectionList` from  https://github.com/bitwarden/clients/pull/3651 for use in this `AccessSelectorComponent` so the component can get reviewed/merged before the Groups tab PRs are reviewed/merged.

The ngSelect component (https://github.com/bitwarden/clients/pull/3383) still needs to be fully integrated before this is functionally complete.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

- **apps/web/src/app/organizations/components/access-selector/*:** The new `AccessSelector` component and module that includes associated models/views.
- **apps/web/src/app/organizations/organization.module.ts:** Add a feature level organization module so that the new `AccessSelectorComponent` can be imported closer to the feature, instead of globally in the `LooseComponentsModule`
- **libs/angular/src/utils/FormSelectionList.ts:** See https://github.com/bitwarden/clients/pull/3651 for more details
- **apps/web/src/locales/en/messages.json:** Add some required i18n keys for the component

## Screenshots

![image](https://user-images.githubusercontent.com/8764515/194466630-cb6fbd14-cec7-4658-86bf-241e338c8064.png)

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

Hinton commented:
Please don't add components to a root components directory. https://adr.bitwarden.com/decisions/0011-angular-folder-structure

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.html:
```diff
@@ -0,0 +1,93 @@
+    <bit-label>{{ selectorLabelText }}</bit-label>
+    <!--TODO: Swap with CL NgSelect-->
+    <select bitInput (change)="selectItem($event)" (blur)="handleOnTouch()" [disabled]="disabled">
+      <option disabled selected value="">-- Type to filter not implemented yet --</option>
```
**thought:** Interesting, I wonder how the multi select component handles placeholders, haven't tested

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.ts:
```diff
@@ -0,0 +1,212 @@
+  });
+
+  protected itemType = AccessItemType;
+  protected permissionList = [
+    { perm: CollectionPermission.VIEW, labelId: "canView" },
+    { perm: CollectionPermission.VIEW_EXCEPT_PASSWORDS, labelId: "canViewExceptPass" },
+    { perm: CollectionPermission.EDIT, labelId: "canEdit" },
+    { perm: CollectionPermission.EDIT_EXCEPT_PASSWORDS, labelId: "canEditExceptPass" },
+  ];
+  protected initialPermission = CollectionPermission.VIEW;
```
**question(non-blocking):** how come we don't have inputs for this? I guess we have the same permissions for all lists? Would there be a point in making this dynamic now? Or should we just wait until we actually need it?

shane-melton replied:
Yep, didn't feel inputs were necessary for now, keeping this fairly specific to the OAVR changes to keep things _"simple"_.

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.models.ts:
```diff
@@ -0,0 +1,108 @@
+      id: string;
+
+      /**
+       * Text that be displayed in the ngSelect list items
```
**nitpick** Text that **will** be displayed in the ngSelect list items

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.models.ts:
```diff
@@ -0,0 +1,108 @@
+ * for the dropdown in the AccessSelectorComponent
+ * @param value
+ */
+export const convertToPermission = (value: SelectionReadOnlyResponse) => {
+  if (value.readOnly) {
+    return value.hidePasswords
+      ? CollectionPermission.VIEW_EXCEPT_PASSWORDS
+      : CollectionPermission.VIEW;
+  } else {
+    return value.hidePasswords
+      ? CollectionPermission.EDIT_EXCEPT_PASSWORDS
+      : CollectionPermission.EDIT;
+  }
+};
```
**cont.question:** Ah I think I understand why we have fixed permissions. We're actually modifying how the component renders and behaves depending on the permissions assigned to the different values.

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.models.ts:
```diff
@@ -0,0 +1,108 @@
+  }
+};
+
+/**
+ * Converts an AccessItemValue back into a SelectionReadOnly class using the CollectionPermission
+ * to determine the values for `readOnly` and `hidePassword`
+ * @param value
+ */
+export const convertToSelectionReadOnly = (value: AccessItemValue) => {
+  return new SelectionReadOnlyRequest(
+    value.id,
+    readOnly(value.permission),
+    hidePassword(value.permission)
+  );
+};
```
**thought:** I'm wondering if this shouldn't be the parent components responsibility. On the other, I guess moving it would result in a lot of duplicate logic. Maybe that's an indication that this component encapsulates something that's a lot more central to our domain that what I originally thought. We have 3 different parent components but I think they are all actually showing the same thing but from different perspectives, making it beneficial to store the core logic behind those perspectives in the same place - this component. Time will tell I guess, either way, nice job :)

coroiu reviewed (commented):
When we initially discussed this component we were unsure if there would be any point in making it since there would be so many details handling the different entities, making this component just a shell depending on it's parent component for almost everything.

Instead I see that this component has turned into a much more complex component that seems to perfectly matching a central part of our domain regarding permissions handling. As long as we make sure to keep an eye on this component in the future so that it doesn't bloat I think this will be a perfect example of a generalized web/app-specific component (whatever we want to call them)! 

You can sort of see how the realization above sloowly dawned on me in the review comments 😅 

I'm gonna leave out the approval until we get this feature complete, but good work!

coroiu commented:
Also a small design comment, can't have a review without those! 😁 

We have a slight misalignment due to mismatch of paddings, see comparison with figma 

This PR
![image](https://user-images.githubusercontent.com/2285588/194820850-a51a3b1b-0227-40ba-ac7e-7fbb3ab2fe48.png)

Figma
![image](https://user-images.githubusercontent.com/2285588/194820819-5e8667c6-e6e7-4f59-a4b3-6a47c7f988e1.png)

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.html:
```diff
@@ -0,0 +1,93 @@
+
+      <td bitCell *ngIf="usePermissions">
+        <label class="sr-only" [for]="'permission' + i">{{ "permission" | i18n }}</label>
+        <select
```
**suggestion:** don't forget to look through the documentation and double check with design, eg. this select is documented as having the `aria-haspopup` attribute, but it is implemented using the html `select` element so I'm guessing that might not be necessary

shane-melton replied:
Could you point me in the right direction as to where you found the documentation for `aria-haspopup`? I can't seem to find it and/or may be looking in the wrong place.

shane-melton commented:
> Instead I see that this component has turned into a much more complex component that seems to perfectly match a central part of our domain regarding permissions handling. 

Yeah, as briefly discussed during our sync, I realized during development that creating a more generic component would be overly complicated, and we'd still end up duplicating a lot of the same logic in multiple places in the OAVR. So, I went in the direction of making it fairly specific (but still re-usable) for the 3-4 scenarios currently in the OAVR.

>As long as we make sure to keep an eye on this component in the future so that it doesn't bloat I think this will be a perfect example of a generalized web/app-specific component (whatever we want to call them)!

Agreed! I think if we find ourselves starting to bloat this component in future work, it should likely be further broken up at that time. Possibly something like a new `GenericSelectorComponent` that the `AccessSelectorComponent` would be composed of. But that level of abstraction didn't seem necessary at this time (I don't see any other uses yet) and would have made this more complicated. 

> You can sort of see how the realization above sloowly dawned on me in the review comments 😅

Haha no worries, I could have been more descriptive in the PR description

> I'm gonna leave out the approval until we get this feature complete, but good work!

Sounds good! Looks like [ngSelect integration](https://github.com/bitwarden/clients/pull/3383) is getting merged today! 🥳

shane-melton commented:
> Also a small design comment, can't have a review without those! 😁
> 
> We have a slight misalignment due to mismatch of paddings, see comparison with figma

Are you referring to the space between the down arrow for the `<select/>` element and the options label? Or padding between rows of the table? 

I believe that arrow's appearance is controlled differently by each browser, so I'm not sure we can modify that with just CSS. I think it would require a custom _dropdown_ component that would essentially simulate the same functionality as a `<select/>` but with greater control over appearance. We might need another CL component for that, as I see anything like that currently.

coroiu commented:
> Are you referring to the space between the down arrow for the <select/> element and the options label? Or padding between rows of the table?

@shane-melton 
Neither, sorry for being unclear! In the figma design the header text aligns with the text inside the select components. 

![image](https://user-images.githubusercontent.com/2285588/195030800-5b47264f-3374-4cd2-b6fc-8f3d2434851e.png)

There is a small space between the red line and the select text here but I'm not sure if it's on purpose @bitwarden/dept-design ?
![image](https://user-images.githubusercontent.com/2285588/195030868-4b9a7c13-c595-44d2-bee5-0b96e414c11d.png)

Here's another example 
![image](https://user-images.githubusercontent.com/2285588/195031140-20c73703-5a33-435b-a98b-f76878bbe3c8.png)

danielleflinn commented:
@coroiu the small space is unintentional (should be fixed now in Figma). You are correct though in calling out that the header should be aligned with the permission text.

shane-melton commented:
@danielleflinn 

> Would it work to use the [component library link](https://components.bitwarden.com/?path=/story/component-library-link--buttons) for the permission menu? That might improve the spacing between the arrow and selected permission text.

That would help with the spacing between the arrow and text, but we can't use the link directive for rendering the `<select/>` element (which is necessary to get the "selection" form logic). Unfortunately, I believe the only way to make the in-line table permission selector look exactly like the Figma design would be to create a new form component that simulates a `<select/>` tag using popup menus similar to how the ngSelect library does it (we could possibly even re-use that, but I'm not currently familiar with how customizable it is).

I think it would be a good idea to have a customizable select component in our CL, but I'm a little hesitant to try and get one written for this component/PR as this is a blocker for quite a few other OAVR stories. If we're okay with leaving this as is, I think it would be best to create a CL ticket to update (or add to) the [existing select form component](https://components.bitwarden.com/?path=/story/component-library-form-field--select) to support the styles in Figma for this access selector. What do you think?

> @coroiu the small space is unintentional (should be fixed now in Figma). You are correct though in calling out that the header should be aligned with the permission text.

Fortunately, I should be able to fix this easy enough

danielleflinn commented:
@shane-melton sounds good!

danielleflinn commented:
On another note; the avatars in the screenshot look a bit large. In Figma I think I have the small variant. What size are you using here?

shane-melton commented:
> On another note; the avatars in the screenshot look a bit large. In Figma I think I have the small variant. What size are you using here?

Yep, it was using the default size and needed to be set to small. I also adjusted the fonts as they were off by one size as well. Looks much better now, good catch!

shane-melton commented:
Okay, this should be ready to go now for review. 

@vincentsalucci Integration with the multi select component went smoothly! It only required one minor tweak to make the `SelectItemView.ts` `icon` and `parentGrouping` properties nullable. I didn't see that being an issue anywhere, but wanted to mention it in case you see any potential problems with that. [Specific commit details found here.](https://github.com/bitwarden/clients/pull/3717/commits/27e4ab25c1b708afb6fcfe1b9a0d6c728050ee44#diff-74c06f9efe30161aeeae6419962acc9c1dabf070d936fa82d4d3ca340b2eaf79)

I also updated the component stories to include more scenarios show in the design doc (readonly permissions, duplicate groups, overriding groups, access all user/group display, etc.). 

And finally I added some initial unit tests to help add some confidence to its implementation as this is a fairly _"smart"_ component and implements the `ControlValueAccessor` interface.

danielleflinn commented:
@shane-melton it would be more accessible if VoiceOver could announce the data in the corresponding 'name' column along with the 'permission' label when the select is focused. 

Similarly when the Remove button is in focus, it would be ideal for VoiceOver to also announce the corresponding data so the SR user knows which item is being removed. 

That said, I missed calling this out in Figma so I'm fine if we approve this as is, and I can capture the desired changes in a separate jira issue.

shane-melton commented:
> @shane-melton it would be more accessible if VoiceOver could announce the data in the corresponding 'name' column along with the 'permission' label when the select is focused.

@danielleflinn Good call! It was easy to add so I went ahead and put it in there, thanks for pointing it out!

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.spec.ts:
```diff
@@ -0,0 +1,140 @@
+      expect(emptyTableCell?.textContent).toEqual("Nothing selected");
+    });
+
+    it("should show one selected value", () => {
```
suggestion: clarify state under test should show one selected value when what?

shane-melton replied:
I went through and cleaned up all the test states to be more clear

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.spec.ts:
```diff
@@ -0,0 +1,140 @@
+      expect(firstColSpan.textContent).toEqual("Group 1");
+    });
+
+    it("should emit when the internal value changes", () => {
+      const mockChange = jest.fn();
+      let expectedCalls = 0;
+      component.registerOnChange(mockChange);
+      component.permissionMode = "edit";
+
+      // Simulate selecting an item
+      component.selectItems([{ id: "123" } as any]);
+      expectedCalls++;
+
+      expect(mockChange.mock.calls.length).toEqual(expectedCalls);
+      expect(mockChange.mock.lastCall[0]).toHaveProperty("[0].id", "123");
+
+      // Simulate modifying that item
+      component.changeSelectedItemPerm(0, CollectionPermission.EDIT);
+      expectedCalls++;
+
+      expect(mockChange.mock.calls.length).toEqual(expectedCalls);
+      expect(mockChange.mock.lastCall[0]).toHaveProperty("[0].id", "123");
+      expect(mockChange.mock.lastCall[0]).toHaveProperty(
+        "[0].permission",
+        CollectionPermission.EDIT
+      );
+
+      // Simulate deselecting that item
+      component.deselectItem("123");
+      expectedCalls++;
+
+      expect(mockChange.mock.calls.length).toEqual(expectedCalls);
+      expect(mockChange.mock.lastCall[0].length).toEqual(0);
+    });
```
thought: I'm usually very wary of using more than one assertion block because it usually means that the test is testing more than one thing. I'm not gonna add any suggestion instead I just ask that you take a second to reconsider if this can be written in another way

shane-melton replied:
Totally fair, it was definitely testing multiple things in the same test. I split it up to be more explicit about exactly what piece is being tested.

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.ts:
```diff
@@ -12,6 +13,13 @@ import {
+ * - 'readonly = No perm controls. Column rendered an if available on an item. No perm values are emitted
+ * - 'edit' = Perm Controls and column present. Perm values are emitted.
+ */
+type PermissionMode = "hidden" | "readonly" | "edit";
```
suggestion(non-blocking): This could be written as an enum

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.ts:
```diff
@@ -27,18 +35,33 @@ export class AccessSelectorComponent implements ControlValueAccessor, OnInit, On
    */
   protected selectionList = new FormSelectionList<AccessItemView, AccessItemValue>((item) => {
-    return this.formBuilder.group({
+    const permControl = this.formBuilder.control(this.initialPermission);
```
nit: took me a second to figure out what `permControl` meant, I don't think we're gaining anything from shortening it, just write it out :D

shane-melton replied:
Fair enough haha

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.ts:
```diff
@@ -69,25 +92,38 @@ export class AccessSelectorComponent implements ControlValueAccessor, OnInit, On
+    );
+    this.selectionList.populateItems(
+      val.map((m) => {
+        m.icon = m.icon ?? this.itemIcon(m); // Ensure an icon is set
```
thought: should this really be done inside of a getter, they are usually idempotent

shane-melton replied:
I could move it to `ngOnChanges`, but it seemed more legible to do in the `setter` as it is responsible for keeping the internal selection list up to date with the `items` passed into the component. (i.e. if those items change via a parent, we need to immediately update the selection list to make sure it shows the correct items in the template)

coroiu replied:
Oh noo my bad, this is a setter, I thought it was a getter! No this in the correct place! Forget I said anything 😅

coroiu commented on libs/angular/src/utils/FormSelectionList.ts:
```diff
@@ -0,0 +1,201 @@
+ * property for identification/comparison. The second type parameter TControlValue represents the value
+ * type of the form control.
+ */
+export class FormSelectionList<
```
issue: This file should probably have name in kebab-case `form-selection-list`

coroiu commented on libs/angular/src/utils/FormSelectionList.ts:
```diff
@@ -0,0 +1,201 @@
+ * property for identification/comparison. The second type parameter TControlValue represents the value
+ * type of the form control.
+ */
+export class FormSelectionList<
```
suggestion: I think I've mentioned this in a comment somewhere but I really think file should be tested since it plays a central role while also being hard to understand due to being so generic

shane-melton replied:
Yep totally agree, I'll work on getting some tests added right away

shane-melton commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.html:
```diff
@@ -30,59 +32,107 @@
+      <td bitCell *ngIf="permissionMode != 'hidden'">
+        <ng-container
+          *ngIf="
+            permissionMode == 'edit' && !item.readonly && !item.accessAllItems;
+            else readOnlyPerm
```
Good point, it is fairly ugly to keep complex boolean logic in the template. Instead of a getter I moved the logic to a named function `canEditItemPermission(item)` so that it is much easier to read in the template.

Also, I attempted to extract the row into its own separate component, but it did unfortunately start breaking the form group stuff. Angular doesn't seem to like form groups / arrays being split across components it seems, at least not without a lot of hacky work.

coroiu replied:
Ok then I think the function is a good compromise! Thanks for trying! I'll mark this as resolved!

shane-melton commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.spec.ts:
```diff
@@ -0,0 +1,140 @@
+      fixture.detectChanges();
+    });
+
+    it("should have no items selected initially", () => {
```
Done

shane-melton commented:
@coroiu I believe I've addressed all your comments, thanks for the feedback! Let me know if you see anything else!

coroiu commented on apps/web/src/app/organizations/components/access-selector/access-selector.component.ts:
```diff
@@ -13,12 +13,22 @@ import {
+  /**
+   * No permission controls or column present. No permission values are emitted.
+   */
+  HIDDEN = "hidden",
```
nit: I think Pascal case is the most common https://www.typescriptlang.org/docs/handbook/enums.html
I'm not sure if we at bitwarden have another convention though

Sorry for commenting on this again haha
I'll mark it as non-blocking and leave it up to you, it seems silly to block the PR over this.
Maybe you can get someone to approve it on my behalf!

coroiu replied:
Ah, just noticed build errors so my approval is gonna get dismissed either way 🤷

coroiu commented on libs/angular/src/utils/form-selection-list.spec.ts:
```diff
@@ -0,0 +1,253 @@
+];
+const totalTestItemCount = initialTestItems.length;
+
+describe("FormSelectionList", () => {
```
Top notch tests, and I see some unhappy path tests as well, I have 0 comments!  😁

coroiu reviewed (approved):
Everything fixed and looking good!

I never imagined that the test naming convention suggestion during the retro would be used in practice so quickly, but I love it! I promise that the suggestions wasn't related to this PR, they've been on my mind a while!

Quick comment regarding the AAA structure: 
I (personally, this is a non-blocking comment) usually don't write out the section names eg. `//Arrange`. Instead I just split up the sections using newlines, usually it's pretty obvious which part is which. If the test case gets very big and the newlines are needed for organizing within a section, then I usually consider explicitly writing out the test section names. But because your were already using the AAA structure (intentionally or otherwise) I honestly didn't feel the need to adjust it. You already had good tests and the new form selection list tests are spot-on.
